packages/upload-api/src/blob/add.js

```diff
@@ -0,0 +1,153 @@
+import * as Server from '@ucanto/server'
+import { Message } from '@ucanto/core'
+import { ed25519 } from '@ucanto/principal'
+import { CAR } from '@ucanto/transport'
+import * as Blob from '@web3-storage/capabilities/blob'
+import * as UCAN from '@web3-storage/capabilities/ucan'
+import * as API from '../types.js'
+
+import { BlobItemSizeExceeded } from './lib.js'
+
+/**
+ * @param {API.BlobServiceContext} context
+ * @returns {API.ServiceMethod<API.BlobAdd, API.BlobAddSuccess, API.BlobAddFailure>}
+ */
+export function blobAddProvider(context) {
+  return Server.provideAdvanced({
+    capability: Blob.add,
+    handler: async ({ capability, invocation }) => {
+      const {
+        id,
+        allocationsStorage,
+        maxUploadSize,
+        getServiceConnection,
+        tasksStorage,
+      } = context
+      const { blob } = capability.nb
+      const space = /** @type {import('@ucanto/interface').DIDKey} */ (
+        Server.DID.parse(capability.with).did()
+      )
+
+      if (blob.size > maxUploadSize) {
+        return {
+          error: new BlobItemSizeExceeded(maxUploadSize),
+        }
+      }
+
+      const putSubject = await ed25519.derive(blob.content.slice(0, 32))
+      const facts = Object.entries(putSubject.toArchive().keys).map(
+        ([key, value]) => ({
+          did: key,
+          bytes: value,
+        })
+      )
+
+      // Create effects for receipt
+      const blobAllocate = Blob.allocate.invoke({
+        issuer: id,
+        audience: id,
+        with: id.did(),
+        nb: {
+          blob,
+          cause: invocation.link(),
+          space,
+        },
+        expiration: Infinity,
+      })
+      const blobPut = Blob.put.invoke({
+        issuer: putSubject,
+        audience: putSubject,
+        with: putSubject.toDIDKey(),
+        nb: {
+          content: blob.content,
```

Comment:
> @Gozala I thought this was future syntax that represents what we do today with effects. 

Awaits were part of the invocation pipeline, however they only supported `ok` / `error` branches. We have not implemented support for those and in new version we're making `await` more general to allow specifying deeper path. Since we have not implemented old version of await I suggest we go ahead with new syntax, but it's ok if you think we should do the old one instead.

> So, we should add this to the capability definitions right?

Correct we need to add it to the capability definition otherwise fields will be stripped out on encode.

> are they basicaly the same, shouldn't I just want the allocate?

I think you are right we already have dependency with `address` so we should not need another one in form of `_allocate`.